opm/common/utility/CSRGraphFromCoordinates_impl.hpp

```diff
@@ -0,0 +1,546 @@
+    const auto thisNumCols = std::max(this->numCols_, maxColIdx + 1);
+
+    this->preparePushbackRowGrouping(thisNumRows, i);
+
+    this->groupAndTrackColumnIndicesByRow(i, j);
+
+    this->numRows_ = thisNumRows;
+    this->numCols_ = thisNumCols;
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::compress(const Offset maxNumVertices)
+{
+    this->sortColumnIndicesPerRow();
+
+    // Must be called *after* sortColumnIndicesPerRow().
+    this->condenseDuplicates();
+
+    const auto nRows = this->startPointers().size() - 1;
+    if (nRows < maxNumVertices) {
+        this->ia_.insert(this->ia_.end(),
+                         maxNumVertices - nRows,
+                         this->startPointers().back());
+    }
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::sortColumnIndicesPerRow()
+{
+    // Transposition is, in this context, effectively a linear time (O(nnz))
+    // bucket insertion procedure.  In other words transposing the structure
+    // twice creates a structure with column indices in (ascendingly) sorted
+    // order.
+
+    this->transpose();
+    this->transpose();
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::condenseDuplicates()
+{
+    // Note: Must be called *after* sort().
+
+    const auto colIdx = this->ja_;
+    auto end          = colIdx.begin();
+
+    this->ja_.clear();
+
+    [[maybe_unused]] auto compressedIdx = this->compressedIdx_;
+    if constexpr (TrackCompressedIdx) {
+        this->compressedIdx_.clear();
+    }
+
+    const auto numRows = this->ia_.size() - 1;
+    for (auto row = 0*numRows; row < numRows; ++row) {
+        auto begin = end;
+
+        std::advance(end, this->ia_[row + 1] - this->ia_[row + 0]);
+
+        const auto q = this->ja_.size();
+
+        this->condenseAndTrackUniqueColumnsForSingleRow(begin, end);
+
+        this->ia_[row + 0] = q;
+    }
+
+    if constexpr (TrackCompressedIdx) {
+        this->remapCompressedIndex(std::move(compressedIdx));
+    }
+
+    // Record final table sizes.
+    this->ia_.back() = this->ja_.size();
+}
+
+template <typename VertexID, bool TrackCompressedIdx>
+void
+Opm::utility::CSRGraphFromCoordinates<VertexID, TrackCompressedIdx>::
+CSR::preparePushbackRowGrouping(const int         numRows,
+                                const Neighbours& rowIdx)
+{
+    assert (numRows >= 0);
+
+    this->ia_.assign(numRows + 1, 0);
+
+    for (const auto& row : rowIdx) {
+        this->ia_[row + 1] += 1;
+    }
+
+    // Note index range: 1..numRows inclusive.
+    for (typename Start::size_type i = 1, n = numRows; i <= n; ++i) {
+        this->ia_[0] += this->ia_[i];
+        this->ia_[i]  = this->ia_[0] - this->ia_[i];
+    }
+
+    assert (this->ia_[0] == rowIdx.size());
```

Comment:
Why not use ia_.back() instead for this? That would be a more familiar behaviour I think, ia_[0] would usually be just 0.

Reply:
> Why not use `ia_.back()` instead for this

Because at the end of this loop `ia_.back()` does not have that value.  `Ia_.back()` holds the total number of non-zero elements **minus** the number of column indices/neighbours of the vertex corresponding to the last row.

> That would be a more familiar behaviour I think, `ia_[0]` would usually be just 0. That would pop out automatically from `partial_sum()` as well.

Right, and that's why `partial_sum()` does not do the right thing here.  This particular loop uses `ia_[0]` as a crucial accumulator for the total number of edges in the graph.

Reply:
> please document it here and/or near line 386, which is the most magical line of code here

I've pushed an update to expand the reasoning.